circuits/core/src/pedersen/hash.rs

```diff
@@ -0,0 +1,131 @@
+// Copyright (C) 2019-2022 Aleo Systems Inc.
+// This file is part of the snarkVM library.
+
+// The snarkVM library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// The snarkVM library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with the snarkVM library. If not, see <https://www.gnu.org/licenses/>.
+
+use super::Pedersen;
+
+#[cfg(test)]
+use super::WINDOW_SIZE_MULTIPLIER;
+
+use snarkvm_circuits_types::{Boolean, Environment, Group, Inject, Itertools, Ternary, Zero};
+
+use std::borrow::Cow;
+
+impl<E: Environment, const NUM_WINDOWS: usize, const WINDOW_SIZE: usize> Pedersen<E, NUM_WINDOWS, WINDOW_SIZE> {
+    #[allow(dead_code)]
+    fn hash(&self, input: &[Boolean<E>]) -> Group<E> {
+        let constant_false = Boolean::<E>::constant(false);
+
+        let mut input = Cow::Borrowed(input);
+        match input.len() <= WINDOW_SIZE * NUM_WINDOWS {
+            // Pad the input if it is under the required parameter size.
+            true => input.to_mut().resize(WINDOW_SIZE * NUM_WINDOWS, constant_false),
+            // Ensure the input size is within the parameter size,
+            false => E::halt("incorrect input length for pedersen hash"),
+        }
+
+        // Compute sum of h_i^{m_i} for all i.
+        input
+            .chunks(WINDOW_SIZE)
+            .zip_eq(&self.bases)
+            .flat_map(|(bits, powers)| {
+                bits.iter()
+                    .zip_eq(powers)
+                    .map(|(bit, base)| Group::<E>::ternary(bit, &base.clone(), &Group::<E>::zero()))
```

Comment:
Out of curiosity, does `&base.clone()` here need a `.clone()`? I thought pass by reference would be sufficient. (I assume its needed since Clippy CI didn't seem to complain about it)

Reply:
You're absolutely right, I don't need it, will remove it